packages/ubuntu_wsl_setup/lib/pages/profile_setup/profile_setup_model.dart

```diff
@@ -64,10 +69,43 @@ class ProfileSetupModel extends SafeChangeNotifier {
   bool get isValid =>
       realname.isNotEmpty &&
       username.isNotEmpty &&
+      usernameOk &&
       RegExp(kValidUsernamePattern).hasMatch(username) &&
       password.isNotEmpty &&
       password == confirmedPassword;
 
+  /// The server response on whether the desired username is available.
+  UsernameValidation get usernameValidation => _usernameValidation.value;
+  bool get usernameOk => _usernameValidation.value == UsernameValidation.OK;
+
+  Future<void> validate() async {
+    if (username.isNotEmpty &&
+        RegExp(kValidUsernamePattern).hasMatch(username)) {
+      _usernameValidation.value = await _client.validateUsername(username);
+    }
+  }
+
+  final List<FieldValidator<String?>> _validators = [];
+  List<FieldValidator<String?>> get validators => _validators;
+
+  /// Initializes a list of [CallbackValidator]'s for each key-value pair of the
+  /// [errorValueToText] map, in which each value provided as the map key is a
+  /// possible value of [usernameValidation] for which the respective validator
+  /// will return invalid and the mapped value as the validator's error text.
+  void setUsernameValidators(Map<UsernameValidation, String> errorValueToText) {
+    if (_validators.isNotEmpty) {
+      return;
+    }
+    errorValueToText.forEach((key, value) {
+      _validators.add(
+        CallbackValidator(
+          (_) => usernameValidation != key,
+          errorText: value,
+        ),
+      );
+    });
+  }
```

Comment:
I'd propose replacing this with a CallbackValidator placed in the username field widget to avoid widget dependencies in the model.